src/systems/skill/chain_combo_skills.rs

```diff
@@ -0,0 +1,916 @@
+
+		for tree in &trees {
+			update_combos(agent, slots, tree);
+		}
+	}
+}
+
+fn get_trees(
+	combo_trees: &ComboTrees,
+	combo_trees_running: Option<&ComboTreesRunning>,
+	track: &Track<Skill<Active>>,
+) -> Vec<SkillComboTree> {
+	let tree_running = combo_trees_running.and_then(|t| t.0.get(&track.value.data.slot_key));
+	let trees = combo_trees
+		.0
+		.iter()
+		.filter_map(|t| t.get(&track.value.data.slot_key))
+		.collect();
+	let trees = match tree_running {
+		Some(tree) if !tree.next.is_empty() => vec![tree],
+		_ => trees,
+	};
+
+	trees
+		.into_iter()
+		.cloned()
+		.filter(|t| t.skill.name == track.value.name)
+		.collect()
+}
+
+fn clear_combos(agent: &mut EntityCommands, slots: &mut Mut<Slots>) {
+	for slot in &mut slots.0.values_mut() {
+		slot.combo_skill = None;
+	}
+	agent.remove::<ComboTreesRunning>();
+}
+
+fn update_combos(agent: &mut EntityCommands, slots: &mut Mut<Slots>, tree: &SkillComboTree) {
+	let remaining = tree
+		.next
+		.iter()
+		.filter_map(|(slot_key, tree)| update_slot(slots, slot_key, tree));
+
+	agent.insert(ComboTreesRunning(remaining.collect()));
+}
+
+fn update_slot(
+	slots: &mut Mut<Slots>,
+	slot_key: &SlotKey,
+	tree: &SkillComboTree,
+) -> Option<(SlotKey, SkillComboTree)> {
+	let Some(slot) = slots.0.get_mut(slot_key) else {
+		return None;
+	};
+	let Some(item) = &slot.item else {
+		return None;
+	};
+	let skill = &tree.skill;
+	if skill.is_usable_with.intersection(&item.item_type).count() == 0 {
+		return None;
+	};
+	slot.combo_skill = Some(skill.clone());
+	Some((*slot_key, tree.clone()))
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::{
+		components::{Item, ItemType, Side, Slot, SlotKey},
+		skill::SkillComboTree,
+	};
+	use bevy::{
+		app::{App, Update},
+		ecs::entity::Entity,
+		utils::default,
+	};
+	use std::collections::{HashMap, HashSet};
+
+	static SLOTS: [(SlotKey, Slot); 3] = [
+		(
+			SlotKey::Hand(Side::Off),
+			Slot {
+				entity: Entity::from_raw(42),
+				item: None,
+				combo_skill: None,
+			},
+		),
+		(
+			SlotKey::Hand(Side::Main),
+			Slot {
+				entity: Entity::from_raw(43),
+				item: None,
+				combo_skill: None,
+			},
+		),
+		(
+			SlotKey::Legs,
+			Slot {
+				entity: Entity::from_raw(44),
+				item: None,
+				combo_skill: None,
+			},
+		),
+	];
+
+	fn setup(
+		running_skill: &Skill<Active>,
+		combo_trees: &ComboTrees,
+		item_types: &HashMap<SlotKey, HashSet<ItemType>>,
+	) -> (App, Entity) {
+		let track = Track::new(running_skill.clone());
+		let slots = SLOTS.clone().map(|(key, mut slot)| {
+			if let Some(item_type) = item_types.get(&key) {
+				slot.item = Some(Item {
+					item_type: item_type.clone(),
+					..default()
+				});
+			}
+			(key, slot)
+		});
+
+		let mut app = App::new();
+		let agent = app
+			.world
+			.spawn((combo_trees.clone(), track, Slots(HashMap::from(slots))))
+			.id();
+		app.add_systems(Update, chain_combo_skills);
+
+		(app, agent)
+	}
+
+	fn skill_usable_with(types: &HashSet<ItemType>) -> Skill {
```

Comment:
this should maybe just get a slice or array, so we don't have to call `HashSet::from` every time ourselves.

Reply:
Done via local trait. So we can pass different kind of iterables.